# Dem4Cli: check countries included

WCDE provides demographic information for 202 countries/administrative units + the world average, for the year 1950-2100 for three SSP scenarios (ssp1, ssp2, ssp3). Of these, 197 are available in the isipedia country mask.

**1. What countries are included/excluded ?**
   
These 197 countries/administrative units correspond to 184 UN member/observer countries, plus 13 countries/administrations that are not UN sovereign member countries (disputed territories, special adminstrative regions and overseas administrative units, e.g., Hong Kong, French Guyana, Taiwan, Puerto Rico, La Reunion...).

**2. How much global population is omitted?**

Data is matched based on these 197 countries/administrative units. This corresponds to ~99.97% of the global population, (i.e. the missing areas omitted from the analysis only amount to 0.03% of the global population).


In [1]:
import numpy as np
import xarray as xr
import pandas as pd
import geopandas as gpd
import pickle as pk
from scipy import interpolate
import regionmask
import glob, os, re
import openpyxl
import matplotlib.pyplot as plt
import warnings


pd.set_option('display.max_rows', 80)
%matplotlib inline 

from population_demographics import * 

## Part 1. What countries are included/excluded?

### load country metadata

In [3]:
df_metadata = load_country_metadata()
df_metadata

country country_iso3  country_code  \
0                           Afghanistan          AFG             3   
1                               Albania          ALB           103   
2                               Algeria          DZA           203   
3                               Andorra          AND           403   
4                                Angola          AGO           503   
..                                  ...          ...           ...   
190  Venezuela (Bolivarian Republic of)          VEN         22503   
191                            Viet Nam          VNM         22603   
192                               Yemen          YEM         23003   
193                              Zambia          ZMB         23103   
194                            Zimbabwe          ZWE         23203   

                         region         income_group  
0                    South Asia           Low income  
1         Europe & Central Asia  Upper middle income  
2    Middle East & North Africa  Upper middle income  
3         Europe & Central Asia          High income  
4            Sub-Saharan Africa  Lower middle income  
..                          ...                  ...  
190   Latin America & Caribbean                  NaN  
191         East Asia & Pacific  Lower middle income  
192  Middle East & North Africa           Low income  
193          Sub-Saharan Africa  Lower middle income  
194          Sub-Saharan Africa  Lower middle income  

[195 rows x 5 columns]

In [4]:
df_countries_matched = match_country_names_all_mask_frac()

Unmatched ISIMIP countries (without WCDE data) after all merges:
                   country           country_wb
194                Andorra              Andorra
196               Dominica             Dominica
199                  Palau                Palau
7            Liechtenstein        Liechtenstein
8         Marshall Islands     Marshall Islands
9                   Monaco               Monaco
10                   Nauru                Nauru
12   Saint Kitts and Nevis  St. Kitts and Nevis
13              San Marino           San Marino
14                  Tuvalu               Tuvalu
Unmatched WCDE countries after all merges:
                                   country_wcde
2                                         Aruba
3                               Channel Islands
4                                       Curaçao
5  Macao Special Administrative Region of China
6                                         World
Unmatched ISIMIP mask countries (geojson + frac mask) after all merges:
    

In [5]:
df_countries_matched

country   country_wb country_wcde country_mask country_iso3  \
0    Afghanistan  Afghanistan  Afghanistan  Afghanistan          AFG   
1        Albania      Albania      Albania      Albania          ALB   
2        Algeria      Algeria      Algeria      Algeria          DZA   
3        Andorra      Andorra         None      Andorra          AND   
4         Angola       Angola       Angola       Angola          AGO   
..           ...          ...          ...          ...          ...   
228          NaN          NaN          NaN          NaN          NaN   
229          NaN          NaN          NaN          NaN          NaN   
230          NaN          NaN          NaN          NaN          NaN   
231          NaN          NaN          NaN          NaN          NaN   
232          NaN          NaN          NaN          NaN          NaN   

    iso3_mask iso3_frac  country_code                      region  \
0         AFG       AFG           3.0                  South Asia   
1         ALB       ALB         103.0       Europe & Central Asia   
2         DZA       DZA         203.0  Middle East & North Africa   
3         AND       AND         403.0       Europe & Central Asia   
4         AGO       AGO         503.0          Sub-Saharan Africa   
..        ...       ...           ...                         ...   
228       NaN       MNP           NaN                         NaN   
229       NaN      PSID           NaN                         NaN   
230       NaN       SXM           NaN                         NaN   
231       NaN       TCA           NaN                         NaN   
232       NaN       VGB           NaN                         NaN   

            income_group  
0             Low income  
1    Upper middle income  
2    Upper middle income  
3            High income  
4    Lower middle income  
..                   ...  
228                  NaN  
229                  NaN  
230                  NaN  
231                  NaN  
232                  NaN  

[233 rows x 10 columns]

In [79]:
#UN_countries = pd.read_csv('./data-new/ignore/un-countries.csv',sep=',').rename(columns={'country':'country_un'})

# Load UN country list from Wikipedia with sovereignty status (2024 July) 

UN_countries = pd.read_excel('./data-new/ignore/ISO-wikipedia.xlsx').rename(columns={'World Factbook[6]':'country_un','ISO 3166-1-A3':'iso3_un' })[['country_un','Sovereignty','iso3_un']]

In [80]:
UN_countries = UN_countries.dropna(subset=['iso3_un'])
UN_countries

country_un    Sovereignty iso3_un
0               Islamic Republic of Afghanistan      UN member     AFG
1                                         Åland        Finland     ALA
2                           Republic of Albania      UN member     ALB
3       People's Democratic Republic of Algeria      UN member     DZA
4                   Territory of American Samoa  United States     ASM
..                                          ...            ...     ...
269  Territory of the Wallis and Futuna Islands         France     WLF
270            Sahrawi Arab Democratic Republic  Disputed [ak]     ESH
271                           Republic of Yemen      UN member     YEM
272                          Republic of Zambia      UN member     ZMB
273                        Republic of Zimbabwe      UN member     ZWE

[249 rows x 3 columns]

In [81]:
#merge datasets

df_merge = pd.merge(df_countries_matched,UN_countries, left_on='iso3_frac',right_on='iso3_un',how='outer',indicator='merge_iso')
df_merge

country   country_wb country_wcde country_mask country_iso3  \
0    Afghanistan  Afghanistan  Afghanistan  Afghanistan          AFG   
1        Albania      Albania      Albania      Albania          ALB   
2        Algeria      Algeria      Algeria      Algeria          DZA   
3        Andorra      Andorra         None      Andorra          AND   
4         Angola       Angola       Angola       Angola          AGO   
..           ...          ...          ...          ...          ...   
255          NaN          NaN          NaN          NaN          NaN   
256          NaN          NaN          NaN          NaN          NaN   
257          NaN          NaN          NaN          NaN          NaN   
258          NaN          NaN          NaN          NaN          NaN   
259          NaN          NaN          NaN          NaN          NaN   

    iso3_mask iso3_frac  country_code                      region  \
0         AFG       AFG           3.0                  South Asia   
1         ALB       ALB         103.0       Europe & Central Asia   
2         DZA       DZA         203.0  Middle East & North Africa   
3         AND       AND         403.0       Europe & Central Asia   
4         AGO       AGO         503.0          Sub-Saharan Africa   
..        ...       ...           ...                         ...   
255       NaN       NaN           NaN                         NaN   
256       NaN       NaN           NaN                         NaN   
257       NaN       NaN           NaN                         NaN   
258       NaN       NaN           NaN                         NaN   
259       NaN       NaN           NaN                         NaN   

            income_group                                         country_un  \
0             Low income                    Islamic Republic of Afghanistan   
1    Upper middle income                                Republic of Albania   
2    Upper middle income            People's Democratic Republic of Algeria   
3            High income                            Principality of Andorra   
4    Lower middle income                                 Republic of Angola   
..                   ...                                                ...   
255                  NaN                       Collectivity of Saint-Martin   
256                  NaN                             Republic of San Marino   
257                  NaN                                            Tokelau   
258                  NaN  Baker Island, Howland Island, Jarvis Island, J...   
259                  NaN         Territory of the Wallis and Futuna Islands   

       Sovereignty iso3_un   merge_iso  
0        UN member     AFG        both  
1        UN member     ALB        both  
2        UN member     DZA        both  
3        UN member     AND        both  
4        UN member     AGO        both  
..             ...     ...         ...  
255         France     MAF  right_only  
256      UN member     SMR  right_only  
257    New Zealand     TKL  right_only  
258  United States     UMI  right_only  
259         France     WLF  right_only  

[260 rows x 14 columns]

In [82]:
df_merge[df_merge['merge_iso']=='both']


country   country_wb country_wcde country_mask country_iso3  \
0    Afghanistan  Afghanistan  Afghanistan  Afghanistan          AFG   
1        Albania      Albania      Albania      Albania          ALB   
2        Algeria      Algeria      Algeria      Algeria          DZA   
3        Andorra      Andorra         None      Andorra          AND   
4         Angola       Angola       Angola       Angola          AGO   
..           ...          ...          ...          ...          ...   
226          NaN          NaN          NaN          NaN          NaN   
228          NaN          NaN          NaN          NaN          NaN   
230          NaN          NaN          NaN          NaN          NaN   
231          NaN          NaN          NaN          NaN          NaN   
232          NaN          NaN          NaN          NaN          NaN   

    iso3_mask iso3_frac  country_code                      region  \
0         AFG       AFG           3.0                  South Asia   
1         ALB       ALB         103.0       Europe & Central Asia   
2         DZA       DZA         203.0  Middle East & North Africa   
3         AND       AND         403.0       Europe & Central Asia   
4         AGO       AGO         503.0          Sub-Saharan Africa   
..        ...       ...           ...                         ...   
226       NaN       CUW           NaN                         NaN   
228       NaN       MNP           NaN                         NaN   
230       NaN       SXM           NaN                         NaN   
231       NaN       TCA           NaN                         NaN   
232       NaN       VGB           NaN                         NaN   

            income_group                                    country_un  \
0             Low income               Islamic Republic of Afghanistan   
1    Upper middle income                           Republic of Albania   
2    Upper middle income       People's Democratic Republic of Algeria   
3            High income                       Principality of Andorra   
4    Lower middle income                            Republic of Angola   
..                   ...                                           ...   
226                  NaN                            Country of Curaçao   
228                  NaN  Commonwealth of the Northern Mariana Islands   
230                  NaN                                  Sint Maarten   
231                  NaN                      Turks and Caicos Islands   
232                  NaN                                Virgin Islands   

        Sovereignty iso3_un merge_iso  
0         UN member     AFG      both  
1         UN member     ALB      both  
2         UN member     DZA      both  
3         UN member     AND      both  
4         UN member     AGO      both  
..              ...     ...       ...  
226     Netherlands     CUW      both  
228   United States     MNP      both  
230     Netherlands     SXM      both  
231  United Kingdom     TCA      both  
232  United Kingdom     VGB      both  

[222 rows x 14 columns]

In [86]:
unmatched_lx = df_merge[df_merge['merge_iso']=='left_only']
unmatched_lx

country     country_wb  \
100  Liechtenstein  Liechtenstein   
101         Monaco         Monaco   
102     San Marino     San Marino   
103            NaN            NaN   
104            NaN            NaN   
105            NaN            NaN   
106            NaN            NaN   
107            NaN            NaN   
225            NaN            NaN   
227            NaN            NaN   
229            NaN            NaN   

                                     country_wcde country_mask country_iso3  \
100                                          None          NaN          LIE   
101                                          None          NaN          MCO   
102                                          None          NaN          SMR   
103                                         Aruba          NaN          NaN   
104                               Channel Islands          NaN          NaN   
105                                       Curaçao          NaN          NaN   
106  Macao Special Administrative Region of China          NaN          NaN   
107                                         World          NaN          NaN   
225                                           NaN          NaN          NaN   
227                                           NaN          NaN          NaN   
229                                           NaN          NaN          NaN   

    iso3_mask iso3_frac  country_code                 region income_group  \
100       NaN       NaN       11603.0  Europe & Central Asia  High income   
101       NaN       NaN       13503.0  Europe & Central Asia  High income   
102       NaN       NaN       18003.0  Europe & Central Asia  High income   
103       NaN       NaN           NaN                    NaN          NaN   
104       NaN       NaN           NaN                    NaN          NaN   
105       NaN       NaN           NaN                    NaN          NaN   
106       NaN       NaN           NaN                    NaN          NaN   
107       NaN       NaN           NaN                    NaN          NaN   
225       NaN      CSID           NaN                    NaN          NaN   
227       NaN     IOSID           NaN                    NaN          NaN   
229       NaN      PSID           NaN                    NaN          NaN   

    country_un Sovereignty iso3_un  merge_iso  
100        NaN         NaN     NaN  left_only  
101        NaN         NaN     NaN  left_only  
102        NaN         NaN     NaN  left_only  
103        NaN         NaN     NaN  left_only  
104        NaN         NaN     NaN  left_only  
105        NaN         NaN     NaN  left_only  
106        NaN         NaN     NaN  left_only  
107        NaN         NaN     NaN  left_only  
225        NaN         NaN     NaN  left_only  
227        NaN         NaN     NaN  left_only  
229        NaN         NaN     NaN  left_only

In [87]:
unmatched_rx = df_merge[df_merge['merge_iso']=='right_only']
unmatched_rx

country country_wb country_wcde country_mask country_iso3 iso3_mask  \
233     NaN        NaN          NaN          NaN          NaN       NaN   
234     NaN        NaN          NaN          NaN          NaN       NaN   
235     NaN        NaN          NaN          NaN          NaN       NaN   
236     NaN        NaN          NaN          NaN          NaN       NaN   
237     NaN        NaN          NaN          NaN          NaN       NaN   
238     NaN        NaN          NaN          NaN          NaN       NaN   
239     NaN        NaN          NaN          NaN          NaN       NaN   
240     NaN        NaN          NaN          NaN          NaN       NaN   
241     NaN        NaN          NaN          NaN          NaN       NaN   
242     NaN        NaN          NaN          NaN          NaN       NaN   
243     NaN        NaN          NaN          NaN          NaN       NaN   
244     NaN        NaN          NaN          NaN          NaN       NaN   
245     NaN        NaN          NaN          NaN          NaN       NaN   
246     NaN        NaN          NaN          NaN          NaN       NaN   
247     NaN        NaN          NaN          NaN          NaN       NaN   
248     NaN        NaN          NaN          NaN          NaN       NaN   
249     NaN        NaN          NaN          NaN          NaN       NaN   
250     NaN        NaN          NaN          NaN          NaN       NaN   
251     NaN        NaN          NaN          NaN          NaN       NaN   
252     NaN        NaN          NaN          NaN          NaN       NaN   
253     NaN        NaN          NaN          NaN          NaN       NaN   
254     NaN        NaN          NaN          NaN          NaN       NaN   
255     NaN        NaN          NaN          NaN          NaN       NaN   
256     NaN        NaN          NaN          NaN          NaN       NaN   
257     NaN        NaN          NaN          NaN          NaN       NaN   
258     NaN        NaN          NaN          NaN          NaN       NaN   
259     NaN        NaN          NaN          NaN          NaN       NaN   

    iso3_frac  country_code region income_group  \
233       NaN           NaN    NaN          NaN   
234       NaN           NaN    NaN          NaN   
235       NaN           NaN    NaN          NaN   
236       NaN           NaN    NaN          NaN   
237       NaN           NaN    NaN          NaN   
238       NaN           NaN    NaN          NaN   
239       NaN           NaN    NaN          NaN   
240       NaN           NaN    NaN          NaN   
241       NaN           NaN    NaN          NaN   
242       NaN           NaN    NaN          NaN   
243       NaN           NaN    NaN          NaN   
244       NaN           NaN    NaN          NaN   
245       NaN           NaN    NaN          NaN   
246       NaN           NaN    NaN          NaN   
247       NaN           NaN    NaN          NaN   
248       NaN           NaN    NaN          NaN   
249       NaN           NaN    NaN          NaN   
250       NaN           NaN    NaN          NaN   
251       NaN           NaN    NaN          NaN   
252       NaN           NaN    NaN          NaN   
253       NaN           NaN    NaN          NaN   
254       NaN           NaN    NaN          NaN   
255       NaN           NaN    NaN          NaN   
256       NaN           NaN    NaN          NaN   
257       NaN           NaN    NaN          NaN   
258       NaN           NaN    NaN          NaN   
259       NaN           NaN    NaN          NaN   

                                            country_un       Sovereignty  \
233                                              Åland           Finland   
234                                           Anguilla    United Kingdom   
235  Antarctica (all land and ice shelves south of ...  Antarctic Treaty   
236                                              Aruba       Netherlands   
237                   Bonaire, Sint Eustatius and Saba       Netherlands   
238               

In [92]:
second_merge = pd.merge(unmatched_lx.iloc[:,0:9], unmatched_rx.iloc[:,9:], left_on='country_iso3', right_on='iso3_un',how='outer',indicator='merge_iso2')
second_merge

country     country_wb  \
0   Liechtenstein  Liechtenstein   
1          Monaco         Monaco   
2      San Marino     San Marino   
3             NaN            NaN   
4             NaN            NaN   
5             NaN            NaN   
6             NaN            NaN   
7             NaN            NaN   
8             NaN            NaN   
9             NaN            NaN   
10            NaN            NaN   
11            NaN            NaN   
12            NaN            NaN   
13            NaN            NaN   
14            NaN            NaN   
15            NaN            NaN   
16            NaN            NaN   
17            NaN            NaN   
18            NaN            NaN   
19            NaN            NaN   
20            NaN            NaN   
21            NaN            NaN   
22            NaN            NaN   
23            NaN            NaN   
24            NaN            NaN   
25            NaN            NaN   
26            NaN            NaN   
27            NaN            NaN   
28            NaN            NaN   
29            NaN            NaN   
30            NaN            NaN   
31            NaN            NaN   
32            NaN            NaN   
33            NaN            NaN   
34            NaN            NaN   

                                    country_wcde country_mask country_iso3  \
0                                           None          NaN          LIE   
1                                           None          NaN          MCO   
2                                           None          NaN          SMR   
3                                          Aruba          NaN          NaN   
4                                Channel Islands          NaN          NaN   
5                                        Curaçao          NaN          NaN   
6   Macao Special Administrative Region of China          NaN          NaN   
7                                          World          NaN          NaN   
8                                            NaN          NaN          NaN   
9                                            NaN          NaN          NaN   
10                                           NaN          NaN          NaN   
11                                           NaN          NaN          NaN   
12                                           NaN          NaN          NaN   
13                                           NaN          NaN          NaN   
14                                           NaN          NaN          NaN   
15                                           NaN          NaN          NaN   
16                                           NaN          NaN          NaN   
17                                           NaN          NaN          NaN   
18                                           NaN          NaN          NaN   
19                                           NaN          NaN          NaN   
20                                           NaN          NaN          NaN   
21                                           NaN          NaN          NaN   
22                                           NaN          NaN          NaN   
23                                           NaN          NaN          NaN   
24                                           NaN          NaN          NaN   
25                                           NaN          NaN          NaN   
26                                           NaN          NaN          NaN   
27                                           NaN          NaN          NaN   
28                                           NaN          NaN          NaN   
29                                           NaN          NaN          NaN   
30                                           NaN          NaN          NaN   
31                                           NaN          NaN          NaN   
32                                           NaN          NaN          NaN   
33                                           NaN          NaN

In [112]:
df_combined = pd.concat([df_merge[df_merge['merge_iso']=='both'],second_merge]).reset_index(drop=True)
df_combined

country   country_wb country_wcde country_mask country_iso3  \
0    Afghanistan  Afghanistan  Afghanistan  Afghanistan          AFG   
1        Albania      Albania      Albania      Albania          ALB   
2        Algeria      Algeria      Algeria      Algeria          DZA   
3        Andorra      Andorra         None      Andorra          AND   
4         Angola       Angola       Angola       Angola          AGO   
..           ...          ...          ...          ...          ...   
252          NaN          NaN          NaN          NaN          NaN   
253          NaN          NaN          NaN          NaN          NaN   
254          NaN          NaN          NaN          NaN          NaN   
255          NaN          NaN          NaN          NaN          NaN   
256          NaN          NaN          NaN          NaN          NaN   

    iso3_mask iso3_frac  country_code                      region  \
0         AFG       AFG           3.0                  South Asia   
1         ALB       ALB         103.0       Europe & Central Asia   
2         DZA       DZA         203.0  Middle East & North Africa   
3         AND       AND         403.0       Europe & Central Asia   
4         AGO       AGO         503.0          Sub-Saharan Africa   
..        ...       ...           ...                         ...   
252       NaN       NaN           NaN                         NaN   
253       NaN       NaN           NaN                         NaN   
254       NaN       NaN           NaN                         NaN   
255       NaN       NaN           NaN                         NaN   
256       NaN       NaN           NaN                         NaN   

            income_group                                         country_un  \
0             Low income                    Islamic Republic of Afghanistan   
1    Upper middle income                                Republic of Albania   
2    Upper middle income            People's Democratic Republic of Algeria   
3            High income                            Principality of Andorra   
4    Lower middle income                                 Republic of Angola   
..                   ...                                                ...   
252                  NaN       Saint Helena, Ascension and Tristan da Cunha   
253                  NaN                       Collectivity of Saint-Martin   
254                  NaN                                            Tokelau   
255                  NaN  Baker Island, Howland Island, Jarvis Island, J...   
256                  NaN         Territory of the Wallis and Futuna Islands   

        Sovereignty iso3_un   merge_iso  merge_iso2  
0         UN member     AFG        both         NaN  
1         UN member     ALB        both         NaN  
2         UN member     DZA        both         NaN  
3         UN member     AND        both         NaN  
4         UN member     AGO        both         NaN  
..              ...     ...         ...         ...  
252  United Kingdom     SHN  right_only  right_only  
253          France     MAF  right_only  right_only  
254     New Zealand     TKL  right_only  right_only  
255   United States     UMI  right_only  right_only  
256          France     WLF  right_only  right_only  

[257 rows x 15 columns]

In [113]:
for col in ['iso3_frac', 'country_iso3', 'iso3_un']:
    value_counts = df_combined[col].value_counts()
    filtered_counts = value_counts[value_counts > 1]
    if not filtered_counts.empty:
        print(f"Column: {col}")
        print(filtered_counts)

# no duplicates 

In [115]:
df_combined['Sovereignty'].value_counts()

Sovereignty
UN member           193
France               12
United Kingdom       12
United States         6
Netherlands           4
Australia             4
British Crown         3
New Zealand           3
Denmark               2
Norway                2
China                 2
UN observer           2
Disputed [ak]         1
Disputed [aa]         1
Finland               1
Antarctic Treaty      1
Name: count, dtype: int64

In [114]:
df_combined[df_combined['Sovereignty']=='UN member']

country     country_wb country_wcde country_mask country_iso3  \
0      Afghanistan    Afghanistan  Afghanistan  Afghanistan          AFG   
1          Albania        Albania      Albania      Albania          ALB   
2          Algeria        Algeria      Algeria      Algeria          DZA   
3          Andorra        Andorra         None      Andorra          AND   
4           Angola         Angola       Angola       Angola          AGO   
..             ...            ...          ...          ...          ...   
190         Zambia         Zambia       Zambia       Zambia          ZMB   
191       Zimbabwe       Zimbabwe     Zimbabwe     Zimbabwe          ZWE   
222  Liechtenstein  Liechtenstein         None          NaN          LIE   
223         Monaco         Monaco         None          NaN          MCO   
224     San Marino     San Marino         None          NaN          SMR   

    iso3_mask iso3_frac  country_code                      region  \
0         AFG       AFG           3.0                  South Asia   
1         ALB       ALB         103.0       Europe & Central Asia   
2         DZA       DZA         203.0  Middle East & North Africa   
3         AND       AND         403.0       Europe & Central Asia   
4         AGO       AGO         503.0          Sub-Saharan Africa   
..        ...       ...           ...                         ...   
190       ZMB       ZMB       23103.0          Sub-Saharan Africa   
191       ZWE       ZWE       23203.0          Sub-Saharan Africa   
222       NaN       NaN       11603.0       Europe & Central Asia   
223       NaN       NaN       13503.0       Europe & Central Asia   
224       NaN       NaN       18003.0       Europe & Central Asia   

            income_group                               country_un Sovereignty  \
0             Low income          Islamic Republic of Afghanistan   UN member   
1    Upper middle income                      Republic of Albania   UN member   
2    Upper middle income  People's Democratic Republic of Algeria   UN member   
3            High income                  Principality of Andorra   UN member   
4    Lower middle income                       Republic of Angola   UN member   
..                   ...                                      ...         ...   
190  Lower middle income                       Republic of Zambia   UN member   
191  Lower middle income                     Republic of Zimbabwe   UN member   
222                  NaN            Principality of Liechtenstein   UN member   
223                  NaN                   Principality of Monaco   UN member   
224                  NaN                   Republic of San Marino   UN member   

    iso3_un   merge_iso merge_iso2  
0       AFG        both        NaN  
1       ALB        both        NaN  
2       DZA        both        NaN  
3       AND        both        NaN  
4       AGO        both        NaN  
..      ...         ...        ...  
190     ZMB        both        NaN  
191     ZWE        both        NaN  
222     LIE  right_only       both  
223     MCO  right_only       both  
224     SMR  right_only       both  

[193 rows x 15 columns]

In [116]:
df_combined[df_combined['Sovereignty']=='UN observer']

country          country_wb                    country_wcde  \
132  Palestine, State of  West Bank and Gaza  Occupied Palestinian Territory   
245                  NaN                 NaN                             NaN   

    country_mask country_iso3 iso3_mask iso3_frac  country_code  \
132          NaN          PSE       NaN       PSE       15603.0   
245          NaN          NaN       NaN       NaN           NaN   

                         region         income_group          country_un  \
132  Middle East & North Africa  Lower middle income  State of Palestine   
245                         NaN                  NaN        The Holy See   

     Sovereignty iso3_un   merge_iso  merge_iso2  
132  UN observer     PSE        both         NaN  
245  UN observer     VAT  right_only  right_only

### dem4cli includes a total of 197 countries, 184 un member/observers + 13 others


In [120]:

len(df_combined[~pd.isna(df_combined['country_wcde']) & ~pd.isna(df_combined['iso3_frac']) & ((df_combined['Sovereignty']=='UN member')|(df_combined['Sovereignty']=='UN observer'))])

184

In [121]:
len(df_combined[~pd.isna(df_combined['country_wcde']) & ~pd.isna(df_combined['iso3_frac']) & ((df_combined['Sovereignty']!='UN member')&(df_combined['Sovereignty']!='UN observer'))])

13

In [122]:
184+13

197

In [123]:
df_combined[~pd.isna(df_combined['country_wcde']) & ~pd.isna(df_combined['iso3_frac']) & ((df_combined['Sovereignty']!='UN member')&(df_combined['Sovereignty']!='UN observer'))]

country            country_wb  \
74   Hong Kong, China (SAR)  Hong Kong SAR, China   
192                     NaN                   NaN   
193                     NaN                   NaN   
194                     NaN                   NaN   
195                     NaN                   NaN   
196                     NaN                   NaN   
197                     NaN                   NaN   
198                     NaN                   NaN   
199                     NaN                   NaN   
200                     NaN                   NaN   
201                     NaN                   NaN   
202                     NaN                   NaN   
203                     NaN                   NaN   

                                         country_wcde  \
74   Hong Kong Special Administrative Region of China   
192                                     French Guiana   
193                                  French Polynesia   
194                                        Guadeloupe   
195                                              Guam   
196                                        Martinique   
197                                           Mayotte   
198                                     New Caledonia   
199                                       Puerto Rico   
200                                           Reunion   
201                          Taiwan Province of China   
202                      United States Virgin Islands   
203                                    Western Sahara   

                     country_mask country_iso3 iso3_mask iso3_frac  \
74                      Hong Kong          HKG       HKG       HKG   
192                 French Guiana          NaN       GUF       GUF   
193              French Polynesia          NaN       PYF       PYF   
194                    Guadeloupe          NaN       GLP       GLP   
195                          Guam          NaN       GUM       GUM   
196                    Martinique          NaN       MTQ       MTQ   
197                       Mayotte          NaN       MYT       MYT   
198                 New Caledonia          NaN       NCL       NCL   
199                   Puerto Rico          NaN       PRI       PRI   
200                       Réunion          NaN       REU       REU   
201                        Taiwan          NaN       TWN       TWN   
202  United States Virgin Islands          NaN       VIR       VIR   
203                Western Sahara          NaN       ESH       ESH   

     country_code               region income_group  \
74         8903.0  East Asia & Pacific  High income   
192           NaN                  NaN          NaN   
193           NaN                  NaN          NaN   
194           NaN                  NaN          NaN   
195           NaN                  NaN          NaN   
196           NaN                  NaN          NaN   
197           NaN                  NaN          NaN   
198           NaN                  NaN          NaN   
199           NaN                  NaN          NaN   
200           NaN                  NaN          NaN   
201           NaN                  NaN          NaN   
202           NaN                  NaN          NaN   
203           NaN                  NaN          NaN   

                                            country_un    Sovereignty iso3_un  \
74   Hong Kong Special Administrative Region of Chi...          China     HKG   
192                                             Guyane         France     GUF   
193                                   French Polynesia         France     PYF   
194                                         Guadeloupe         France     GLP   
195                                  Territory of Guam  United States     GUM   
196                                         Martinique         France     MTQ   
197                              Department of Mayotte         France     MYT   
198                                      New Caledonia         France     NCL 

### ISIMIP isipedia-countries 195 countries includes Hong Kong and excludes Vatican

In [124]:
len(df_combined[ ~pd.isna(df_combined['country_iso3']) & ((df_combined['Sovereignty']=='UN member')|(df_combined['Sovereignty']=='UN observer'))])

194

In [125]:
len(df_combined[ ~pd.isna(df_combined['country_iso3'])])

195

In [126]:
df_combined[ ~pd.isna(df_combined['country_iso3']) & (~(df_combined['Sovereignty']=='UN member')& ~(df_combined['Sovereignty']=='UN observer'))]

country            country_wb  \
74  Hong Kong, China (SAR)  Hong Kong SAR, China   

                                        country_wcde country_mask  \
74  Hong Kong Special Administrative Region of China    Hong Kong   

   country_iso3 iso3_mask iso3_frac  country_code               region  \
74          HKG       HKG       HKG        8903.0  East Asia & Pacific   

   income_group                                         country_un  \
74  High income  Hong Kong Special Administrative Region of Chi...   

   Sovereignty iso3_un merge_iso merge_iso2  
74       China     HKG      both        NaN

In [130]:
len(df_combined[(df_combined['Sovereignty']=='UN member')|(df_combined['Sovereignty']=='UN observer')])

195

In [134]:
df_combined[((df_combined['Sovereignty']=='UN member')|(df_combined['Sovereignty']=='UN observer'))&pd.isna(df_combined['country_iso3']) ]



country country_wb country_wcde country_mask country_iso3 iso3_mask  \
245     NaN        NaN          NaN          NaN          NaN       NaN   

    iso3_frac  country_code region income_group    country_un  Sovereignty  \
245       NaN           NaN    NaN          NaN  The Holy See  UN observer   

    iso3_un   merge_iso  merge_iso2  
245     VAT  right_only  right_only

## Part 2. How much population is covered?

More than > 99,9% of the population is covered by the dem4cli processsing. 

In 2000 missing population is 0,03%, in 2020 missing population is 0,05%. 

The missing population is in areas of non-overlap between WCDE data and fractional countrymasks used. 

In [2]:
da_pop_demographics = population_demographics_gridscale_global(startyear=2020,
                                                                    endyear=2021,
                                                                    isimip_round=3, 
                                                                    ssp=3)

opening isimip3 - ssp3
loading country masks
interpolating cohort sizes per country
calculating gridscale demographics


/apps/brussel/RL8/skylake/software/dask/2023.9.2-foss-2023a/lib/python3.11/site-packages/dask/array/core.py:4831: PerformanceWarning: Increasing number of chunks by factor of 11
  result = blockwise(
/apps/brussel/RL8/skylake/software/dask/2023.9.2-foss-2023a/lib/python3.11/site-packages/dask/array/core.py:4831: PerformanceWarning: Increasing number of chunks by factor of 11
  result = blockwise(
/apps/brussel/RL8/skylake/software/dask/2023.9.2-foss-2023a/lib/python3.11/site-packages/dask/array/core.py:4831: PerformanceWarning: Increasing number of chunks by factor of 11
  result = blockwise(
/apps/brussel/RL8/skylake/software/dask/2023.9.2-foss-2023a/lib/python3.11/site-packages/dask/array/core.py:4831: PerformanceWarning: Increasing number of chunks by factor of 11
  result = blockwise(
/apps/brussel/RL8/skylake/software/dask/2023.9.2-foss-2023a/lib/python3.11/site-packages/dask/array/core.py:4831: PerformanceWarning: Increasing number of chunks by factor of 11
  result = blockwise(


In [8]:
da_population = load_population(isimip_round=3,ssp=3, startyear=2020,endyear=2021,).compute()

opening isimip3 - ssp3


In [9]:
(da_population.sel(time=2020) - da_pop_demographics.sel(time=2020).sum(dim=['ages'])).sum() / da_population.sel(time=2020).sum(dim=['lat','lon'])

# in 2020, 0,05% of population missing 

<xarray.DataArray ()>
array(0.00053535)
Coordinates:
    time     int64 2020

In [10]:
da_pop_demographics = population_demographics_gridscale_global(startyear=2000,
                                                                    endyear=2001,
                                                                    isimip_round=3, 
                                                                    ssp=3)

da_population = load_population(isimip_round=3,ssp=3, startyear=2000,endyear=2001,).compute()

opening isimip3 - ssp3
loading country masks
interpolating cohort sizes per country
calculating gridscale demographics


/apps/brussel/RL8/skylake/software/dask/2023.9.2-foss-2023a/lib/python3.11/site-packages/dask/array/core.py:4831: PerformanceWarning: Increasing number of chunks by factor of 11
  result = blockwise(
/apps/brussel/RL8/skylake/software/dask/2023.9.2-foss-2023a/lib/python3.11/site-packages/dask/array/core.py:4831: PerformanceWarning: Increasing number of chunks by factor of 11
  result = blockwise(
/apps/brussel/RL8/skylake/software/dask/2023.9.2-foss-2023a/lib/python3.11/site-packages/dask/array/core.py:4831: PerformanceWarning: Increasing number of chunks by factor of 11
  result = blockwise(
/apps/brussel/RL8/skylake/software/dask/2023.9.2-foss-2023a/lib/python3.11/site-packages/dask/array/core.py:4831: PerformanceWarning: Increasing number of chunks by factor of 11
  result = blockwise(
/apps/brussel/RL8/skylake/software/dask/2023.9.2-foss-2023a/lib/python3.11/site-packages/dask/array/core.py:4831: PerformanceWarning: Increasing number of chunks by factor of 11
  result = blockwise(


opening isimip3 - ssp3


In [11]:
(da_population.sel(time=2000) - da_pop_demographics.sel(time=2000).sum(dim=['ages'])).sum() / da_population.sel(time=2000).sum(dim=['lat','lon'])

# 0,03% in 2000

<xarray.DataArray ()>
array(0.00029752)
Coordinates:
    time     int64 2000